Implement AOS->SOAand SOA->AOS of VW4

CongMa13 opened the pull request:
[==========] Running 20 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 20 tests from KernelTests/TransfromsTest
[ RUN      ] KernelTests/TransfromsTest.RunKernel/0
WSize, DataT, VW, K
w64, i8, 4, 16[       OK ] KernelTests/TransfromsTest.RunKernel/0 (206 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/1
w64, f16, 4, 16[       OK ] KernelTests/TransfromsTest.RunKernel/1 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/2
w64, f32, 4, 16[       OK ] KernelTests/TransfromsTest.RunKernel/2 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/3
w64, f64, 4, 16[       OK ] KernelTests/TransfromsTest.RunKernel/3 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/4
w64, i8, 4, 32[       OK ] KernelTests/TransfromsTest.RunKernel/4 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/5
w64, f16, 4, 32[       OK ] KernelTests/TransfromsTest.RunKernel/5 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/6
w64, f32, 4, 32[       OK ] KernelTests/TransfromsTest.RunKernel/6 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/7
w64, f64, 4, 32[       OK ] KernelTests/TransfromsTest.RunKernel/7 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/8
w64, i8, 4, 64[       OK ] KernelTests/TransfromsTest.RunKernel/8 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/9
w64, f16, 4, 64[       OK ] KernelTests/TransfromsTest.RunKernel/9 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/10
w64, f32, 4, 64[       OK ] KernelTests/TransfromsTest.RunKernel/10 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/11
w64, f64, 4, 64[       OK ] KernelTests/TransfromsTest.RunKernel/11 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/12
w64, i8, 4, 128[       OK ] KernelTests/TransfromsTest.RunKernel/12 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/13
w64, f16, 4, 128[       OK ] KernelTests/TransfromsTest.RunKernel/13 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/14
w64, f32, 4, 128[       OK ] KernelTests/TransfromsTest.RunKernel/14 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/15
w64, f64, 4, 128[       OK ] KernelTests/TransfromsTest.RunKernel/15 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/16
w64, i8, 4, 256[       OK ] KernelTests/TransfromsTest.RunKernel/16 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/17
w64, f16, 4, 256[       OK ] KernelTests/TransfromsTest.RunKernel/17 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/18
w64, f32, 4, 256[       OK ] KernelTests/TransfromsTest.RunKernel/18 (0 ms)
[ RUN      ] KernelTests/TransfromsTest.RunKernel/19
w64, f64, 4, 256[       OK ] KernelTests/TransfromsTest.RunKernel/19 (0 ms)
[----------] 20 tests from KernelTests/TransfromsTest (208 ms total)

[----------] Global test environment tear-down
[==========] 20 tests from 1 test suite ran. (208 ms total)
[  PASSED  ] 20 tests.

CongMa13 committed:
Implement aos_soa_16xk_b32, soa_aos_16xk_b32 of vw=4

- Implement aos_soa_16xk_b32, soa_aos_16xk_b32 of vw=4
- Fix bugs

CongMa13 committed:
Implement aos_soa_32xk_b32 and soa_aos_32xk_b32 of VW = 4

* Implement aos_soa_32xk_b32 and soa_aos_32xk_b32 of VW = 4
* use vector_generator instead of unroll
* Fix bugs

CongMa13 committed:
Implement aos_soa_64k_b32, soa_aos_64k_b32 of VW=4

- Implement aos_soa_64k_b32, soa_aos_64k_b32 of VW=4

CongMa13 committed:
Implement aos_soa_128k_b32, soa_aos_128k_b32 of VW=4

- Implement aos_soa_128k_b32, soa_aos_128k_b32 of VW=4

CongMa13 committed:
Add AosToSoa, SoaToAos structs of VW=4

Move code from aos_soa_***xk_b32, soa_aos_***xk_b32 to structs AosToSoa
and SoaToAos.

CongMa13 committed:
Add AosToSoa<256, 4>, SoaToAos<256, 4> structs of VW=4

- Add AosToSoa<256, 4>, SoaToAos<256, 4> structs of VW=4

CongMa13 committed:
Use new algorithm for aos to soa and soa to aos

- Use unified algorithm for 64k, 128k and 256k aos to soa and soa to aos

CongMa13 committed:
Add unit test for transfroms

cgmillette commented on library/include/rocwmma/internal/dpp_impl.hpp:
```diff
@@ -103,13 +104,20 @@ namespace rocwmma
+                            WriteRowMask, // Mask for affected rows
+                            WriteBankMask, // Mask for affected banks
+                            BoundCtrl); // Fill in 0 on invalid indices
+                        return 0; // discard return value
```
This is a bit misleading to return 0 here. Was there an extra copy or something here we are trying to remove?
src0 is definitely a by-copy argument, but return value elision should be able to avoid the extra copy.

CongMa13 replied:
changed to return the result of cross-lane instructions.

cgmillette commented on library/include/rocwmma/internal/dpp_impl.hpp:
```diff
@@ -103,13 +104,20 @@ namespace rocwmma
-                        WriteRowMask, // Mask for affected rows
-                        WriteBankMask, // Mask for affected banks
-                        BoundCtrl); // Fill in 0 on invalid indices
+                    constexpr size_t VecSize = sizeof(DataT) / sizeof(uint32_t);
```
We might want to have a static_assert here that sizeof(DataT) >= sizeof(uint32_t) and that sizeof(DataT) % sizeof(uint32_t) == 0. This will prevent usage with unpacked types.

CongMa13 replied:
added

cgmillette commented on library/include/rocwmma/internal/dpp_impl.hpp:
```diff
@@ -103,13 +104,20 @@ namespace rocwmma
+                    constexpr size_t VecSize = sizeof(DataT) / sizeof(uint32_t);
+                    auto             dpp     = [](auto&& idx, auto&& v0, auto&& v1) {
+                        constexpr auto i = std::decay_t<decltype(idx)>::value;
+                        *(reinterpret_cast<uint32_t*>(&v0) + i) = __builtin_amdgcn_update_dpp(
```
To make this easier to read, we might try to cast the input vector to uint32_t[VecSize] and index via i

CongMa13 replied:
changed

cgmillette commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -118,7 +118,8 @@ namespace rocwmma
         auto hi    = Blend::Zip2::exec(Dpp::RotateR16<14>::exec(evens), odds);
 
-        return PackUtil::template paddedUnpack<VecSize>(concat(lo, hi));
+        return concat(PackUtil::template paddedUnpack<VecSize / 2u>(lo),
```
Nice catch!

cgmillette commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -118,7 +118,8 @@ namespace rocwmma
         auto lo    = Blend::Zip2::exec(evens, Dpp::RotateR16<2>::exec(odds));
```
Should mark to replace with DPP function

cgmillette commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -162,7 +165,8 @@ namespace rocwmma
         lo          = Blend::Zip16::exec(lo, rot_hi);
```
Mark to replace with dpp::move

CongMa13 replied:
Will do it in the next commit.

cgmillette commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -181,9 +185,88 @@ namespace rocwmma
         lo          = Blend::Zip32::exec(lo, rot_hi);
```
mark to replace with dpp::move

CongMa13 replied:
Added TODO comment

Will do it in the next commit.

cgmillette commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -181,9 +185,88 @@ namespace rocwmma
+        auto unpacked_data = PackUtil::template paddedUnpack<4>(packed_data);
+        auto lo            = PackUtil::paddedPack(extractLo(unpacked_data));
+        auto hi            = PackUtil::paddedPack(extractHi(unpacked_data));
+        auto rot_lo        = Swizzle::RotateR32<16>::exec(lo);
```
Why rotate again here?

CongMa13 replied:
rotated lo, hi separately to avoid concat them. Not sure if concat costs any resource. 

But don't need to worry it anymore. line 198 - 210 are replaced with unpackLohi16().

cgmillette replied:
Concat should be relatively cheap compared to other manipulation

cgmillette commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -181,9 +185,88 @@ namespace rocwmma
+
+        packed_data = Swizzle::RotateR32<16>::exec(packed_data);
+
+        auto unpacked_data = PackUtil::template paddedUnpack<4>(packed_data);
```
Can we use the UnpackLoHi16 function here?

CongMa13 replied:
done

cgmillette commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -527,6 +613,272 @@ namespace rocwmma
+        template <typename DataT>
+        ROCWMMA_DEVICE constexpr static inline auto exec(VecT<DataT, VecSize> const& v)
+        {
+            auto unpacked_data = aos_soa_wave_b32_vw4(v);
```
Recommend to put the wave implementation here (RDNA will be different anyway). Then from the 64+, call on this function

CongMa13 replied:
done

cgmillette commented on library/include/rocwmma/internal/transforms_impl.hpp:
```diff
@@ -527,6 +613,272 @@ namespace rocwmma
+        ROCWMMA_DEVICE constexpr static inline auto exec(VecT<DataT, VecSize> const& v)
+        {
+            // Step 1 :  RE-PACK Banks
+            auto v0 = VecT<DataT, VecSize / 2>{v.data[0], v.data[4], v.data[1], v.data[5]};
```
This is analogous to 
```
v0 = vector_utils::unpackLo(lo(v), hi(v))
v1 = vector_utils::unpackHi(lo(v), hi(v))
```

CongMa13 replied:
done

cgmillette commented on test/unit/transforms_test/test/transforms.cpp:
```diff
@@ -0,0 +1,106 @@
+        {
+            auto warpSize = HipDevice::instance()->warpSize();
+            // clang-format off
+            return { {warpSize, 1} };
```
Should we test multiple waves?

CongMa13 replied:
done

CongMa13 committed:
Return generated value from vector_generator

Return generated value from vector_generator
- DPP
- Swizzle
- Permute

Replaced some duplicated code with functions

cgmillette reviewed (approved):
Looking good!
